### PR TITLE
Geneva action reimage vm

### PR DESCRIPTION
**Pull request by charlesakalugwu**
This work implements a geneva action that allows an admin to reimage a given virtual machine, identified by its hostname, in a cluster.

This is part of AZURE-257

/cc @mjudeikis @Makdaam @y-cote @kwoodson

### CODE REVIEW

File: pkg/api/plugin.go [hunk 1]
```diff
@@ -119,4 +119,8 @@ type Plugin interface {
 
 	// GetControlPlanePods fetches a consolidated list of the control plane pods in the cluster
 	GetControlPlanePods(ctx context.Context, oc *OpenShiftManagedCluster) ([]byte, error)
+
+	// ReimageAndRestart reimages and restarts a virtual machine within a scale set in the cluster.
+	// wipeTempDisks indicates whether temporary disks on the VM should be reimaged/wiped as well.
+	ReimageAndRestart(ctx context.Context, oc *OpenShiftManagedCluster, vmScaleSetName string, instanceId string, wipeTempDisks bool) *PluginError
```
Comment (mjudeikis) [inline pkg/api/plugin.go]: When preparing request you are using `OpenShiftManagedClustersVMReimageAndRestartParameters` and encode them into the payload. But the method does not accept anything like that. So you set parameters, encode them into the payload, but never use them in the fakeRP. I get why, but I think we should be aligned with code we produce and how we use it. And this cries for tests....

File: pkg/cluster/admin_actions.go [hunk 2]
```diff
@@ -0,0 +1,19 @@
...
+import (
+	"context"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+)
+
+func (u *simpleUpgrader) ReimageAndRestart(ctx context.Context, cs *api.OpenShiftManagedCluster, vmScalesetName string, instanceId string, wipeTempDisks bool) *api.PluginError {
+	err := u.vmc.Reimage(ctx, cs.Properties.AzProfile.ResourceGroup, vmScalesetName, instanceId, nil)
+	if err != nil {
+		return &api.PluginError{Err: err, Step: "AdminActionReimageVm"}
```
Comment (jim-minter) [inline pkg/cluster/admin_actions.go]: always use consts for PluginError steps

File: pkg/api/plugin.go (hunk 1 above)
Comment (jim-minter) [reply pkg/api/plugin.go]: remove the wipeTempDisk argument?

Comment (mjudeikis) [reply pkg/api/plugin.go]: Or use the parameters object and parse it inside and decide.

Comment (charlesakalugwu) [reply pkg/api/plugin.go]: the reason for the `wipeTempDisk` parameter here is so I can do this: https://github.com/openshift/openshift-azure/pull/1091/commits/10aa7182735ebd0ae5586ff8b82c874b3b13c2db#diff-643883fea487eafcb41824e842b9fd36R10

File: pkg/api/plugin.go [hunk 3]
```diff
@@ -121,4 +123,7 @@ type GenevaActions interface {
 
 	// GetControlPlanePods fetches a consolidated list of the control plane pods in the cluster
 	GetControlPlanePods(ctx context.Context, oc *OpenShiftManagedCluster) ([]byte, error)
+
+	// ReimageAndRestart reimages and restarts a virtual machine within a scale set in the cluster.
+	ReimageAndRestart(ctx context.Context, oc *OpenShiftManagedCluster, vmScaleSetName string, instanceId string) *PluginError
```
Comment (jim-minter) [inline pkg/api/plugin.go]: should accept a hostname, not vmscalesetname/instanceid

File: test/e2e/specs/fakerp/admin.go [hunk 4]
```diff
@@ -103,4 +103,23 @@ var _ = Describe("Openshift on Azure admin e2e tests [AzureClusterReader][Fake]"
 		By("comparing the update blob before and after an update")
 		Expect(reflect.DeepEqual(before, after)).To(Equal(true))
 	})
+
+	It("should be possible for an SRE to reimage and restart a VM in a scale set", func() {
+		if os.Getenv("RESOURCEGROUP") == "" {
```
Comment (jim-minter) [inline test/e2e/specs/fakerp/admin.go]: ugh - `Expect(os.Getenv("RESOURCEGROUP")).ToNot(Equal(""))`

File: pkg/cluster/admin_actions.go [hunk 5]
```diff
@@ -0,0 +1,19 @@
...
+
+	"github.com/openshift/openshift-azure/pkg/api"
+)
+
+func (u *simpleUpgrader) ReimageAndRestart(ctx context.Context, cs *api.OpenShiftManagedCluster, vmScalesetName string, instanceId string) *api.PluginError {
+	err := u.vmc.Reimage(ctx, cs.Properties.AzProfile.ResourceGroup, vmScalesetName, instanceId, nil)
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepReimageAndRestartVmReimageStep}
+	}
+	err = u.vmc.Restart(ctx, cs.Properties.AzProfile.ResourceGroup, vmScalesetName, instanceId)
```
Comment (jim-minter) [inline pkg/cluster/admin_actions.go]: Is the restart superfluous?  I think Reimage on a running machine also restarts it?

File: test/e2e/specs/fakerp/admin.go [hunk 6]
```diff
@@ -103,4 +103,23 @@ var _ = Describe("Openshift on Azure admin e2e tests [AzureClusterReader][Fake]"
...
+		before, err := cli.OpenShiftManagedClustersAdmin.Get(context.Background(), os.Getenv("RESOURCEGROUP"), os.Getenv("RESOURCEGROUP"))
+		Expect(err).NotTo(HaveOccurred())
+		Expect(before).NotTo(BeNil())
+
+		By("Executing reimage and restart on a vm in the cluster")
+		update, err := cli.OpenShiftManagedClustersAdmin.ReimageAndRestartAndWait(context.Background(), os.Getenv("RESOURCEGROUP"), os.Getenv("RESOURCEGROUP"), "ss-master", "0")
+		Expect(err).NotTo(HaveOccurred())
+		Expect(update.StatusCode).To(Equal(http.StatusOK))
+		Expect(update).NotTo(BeNil())
+	})
```
Comment (jim-minter) [inline test/e2e/specs/fakerp/admin.go]: You need to prove that the instance was reimaged.
You should then run the cluster sanity checks.

File: pkg/api/plugin.go (hunk 3 above)
Comment (jim-minter) [reply pkg/api/plugin.go]: I think you can return a straight `error` here - don't worry about PluginError consts

File: pkg/plugin/plugin.go [hunk 7]
```diff
@@ -240,3 +240,12 @@ func (p *plugin) GetControlPlanePods(ctx context.Context, oc *api.OpenShiftManag
...
 }
+
+func (p *plugin) ReimageAndRestart(ctx context.Context, oc *api.OpenShiftManagedCluster, vmScaleSetName string, instanceId string) *api.PluginError {
+	err := p.clusterUpgrader.CreateClients(ctx, oc, true)
+	if err != nil {
+		return &api.PluginError{Err: err, Step: api.PluginStepClientCreation}
+	}
+	p.log.Infof("reimaging and restarting vm id=%s in %s scaleset", instanceId, vmScaleSetName)
+	return p.clusterUpgrader.ReimageAndRestart(ctx, oc, vmScaleSetName, instanceId)
+}
```
Comment (jim-minter) [inline pkg/plugin/plugin.go]: I think you should wait for node readiness here

File: pkg/cluster/admin_actions.go (hunk 5 above)
Comment (thekad) [reply pkg/cluster/admin_actions.go]: I think it is, the docs https://docs.microsoft.com/en-us/rest/api/compute/virtualmachinescalesetvms/reimage say it upgrades the OS, so in theory after reimage there is a "first boot"

Comment (charlesakalugwu) [reply pkg/cluster/admin_actions.go]: i tried this out in the portal and reimaging a running vm does include a restart. I'll change the entire API from `ReimageAndRestart` to `Reimage`

image

File: pkg/api/plugin.go (hunk 3 above)
Comment (charlesakalugwu) [reply pkg/api/plugin.go]: done

File: pkg/plugin/plugin_test.go [hunk 8]
```diff
@@ -224,3 +226,69 @@ func TestForceUpdate(t *testing.T) {
...
 }
+
+func TestReimage(t *testing.T) {
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	tests := []struct {
+		name     string
+		hostname string
+		isMaster bool
```
Comment (charlesakalugwu) [inline pkg/plugin/plugin_test.go]: this seems weak...

File: pkg/cluster/admin_actions.go [hunk 9]
```diff
@@ -0,0 +1,59 @@
...
+import (
+	"context"
+	"fmt"
+
+	"github.com/openshift/openshift-azure/pkg/api"
+	"github.com/openshift/openshift-azure/pkg/cluster/kubeclient"
+)
+
+// vmNames represents the id of a vm within a scale set, its computer's name and its owning scale set's name
+type vmNames struct {
```
Comment (kwoodson) [inline pkg/cluster/admin_actions.go]: consider: singular name `vmName`

File: pkg/cluster/kubeclient/status.go [hunk 10]
```diff
@@ -8,6 +8,10 @@ import (
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
 )
 
+const (
+	MasterNodeRole = "node-role.kubernetes.io/master"
```
Comment (charlesakalugwu) [inline pkg/cluster/kubeclient/status.go]: perhaps `MasterNodeLabel`

File: test/e2e/specs/fakerp/admin.go (hunk 6 above)
Comment (charlesakalugwu) [reply test/e2e/specs/fakerp/admin.go]: done

File: pkg/fakerp/admin_handlers.go [hunk 11]
```diff
@@ -35,6 +36,28 @@ func (s *Server) handleGetControlPlanePods(w http.ResponseWriter, req *http.Requ
...
 
+// handleReimage handles reimaging a vm in the cluster
+func (s *Server) handleReimage(w http.ResponseWriter, req *http.Request) {
+	cs := s.read()
+	if cs == nil {
+		s.internalError(w, "Failed to read the internal config")
+		return
+	}
+
+	hostname := chi.URLParam(req, "hostname")
```
Comment (charlesakalugwu) [inline pkg/fakerp/admin_handlers.go]: validate `hostname`

File: pkg/cluster/admin_actions.go [hunk 12]
```diff
@@ -0,0 +1,69 @@
...
+
+func (u *simpleUpgrader) ListHostnames(ctx context.Context, resourceGroup string) ([]*VmIdentifier, error) {
+	scaleSets, err := u.ssc.List(ctx, resourceGroup)
+	if err != nil {
+		return nil, err
+	}
+	var names []*VmIdentifier
+	for _, scaleSet := range scaleSets {
+		scalesetName := *scaleSet.Name
+		vms, err := u.vmc.List(ctx, resourceGroup, scalesetName, "", "", "")
```
Comment (kwoodson) [inline pkg/cluster/admin_actions.go]: did you want to expose filter, selectParameter, or expand?

File: pkg/cluster/admin_actions.go [hunk 13]
```diff
@@ -0,0 +1,69 @@
...
+				ScaleSet:   scalesetName,
+			}
+			names = append(names, name)
+		}
+	}
+	return names, nil
+}
+
+func (u *simpleUpgrader) VmIdByHostname(ctx context.Context, resourceGroup string, hostname kubeclient.ComputerName) (*VmIdentifier, error) {
+	vms, err := u.ListHostnames(ctx, resourceGroup)
```
Comment (kwoodson) [inline pkg/cluster/admin_actions.go]: This could take advantage of the hostname field of line #27.

File: test/e2e/specs/fakerp/reimagevm.go [hunk 14]
```diff
@@ -0,0 +1,66 @@
...
+
+		By("Verifying through azure activity logs that the reimage happened")
+		logs, err := azurecli.ActivityLogsByResourceIdAndStatus(context.Background(), *vm.ID, startTime, endTime, azure.ActivitySucceeded)
+		Expect(err).NotTo(HaveOccurred())
+		Expect(logs).NotTo(BeEmpty())
+		loggedOperations := make(map[string]bool)
+		for _, log := range logs {
+			loggedOperations[*log.OperationName.Value] = true
+		}
+		Expect(loggedOperations).To(HaveKey("Microsoft.Compute/virtualMachineScaleSets/virtualmachines/reimage/action"))
```
Comment (kwoodson) [inline test/e2e/specs/fakerp/reimagevm.go]: This is an interesting approach.  Is there anything we can do that verifies the instance?  What if reimage returned `true` but never reimaged the instance.  Just some food for thought.

**Comment by jim-minter** [comment]
/hold
@charlesakalugwu we'll go through this PR in our 1-1, please don't push any changes to it for now.

File: test/e2e/specs/fakerp/reimagevm.go (hunk 14 above)
Comment (jim-minter) [reply test/e2e/specs/fakerp/reimagevm.go]: we expect this operation only to exist if the instance really was reimaged.

**Comment by jim-minter** [comment]
@charlesakalugwu have force pushed to your branch.  Am going to lgtm, if you're happy with it, do a hold cancel.

/lgtm

**Comment by charlesakalugwu** [comment]
@jim-minter I checked your commit locally and everything works fine. I fixed the make verify errors and force-pushed again. Feel free to re-lgtm.

/hold cancel

**Pull request merged**